[SPARK-15370] [SQL] Update RewriteCorrelatedScalarSubquery rule to fix COUNT bug

frreiss opened the pull request:
## What changes were proposed in this pull request?

This pull request fixes the COUNT bug in the `RewriteCorrelatedScalarSubquery` rule.

After this change, the rule tests the expression at the root of the correlated subquery to determine whether the expression returns NULL on empty input. If the expression does not return NULL, the rule generates additional logic in the Project operator above the rewritten subquery.  This additional logic intercepts NULL values coming from the outer join and replaces them with the value that the subquery's expression would return on empty input.
## How was this patch tested?

Added regression tests to cover all branches of the updated rule (see changes to `SubquerySuite.scala`).
Ran all existing automated regression tests after merging with latest trunk.

frreiss committed:
Commit before merge.

frreiss committed:
Merge branch 'master' of https://github.com/apache/spark

frreiss committed:
Merge branch 'master' of https://github.com/apache/spark

frreiss committed:
Merge branch 'master' of https://github.com/apache/spark

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala:
```diff
@@ -293,4 +293,65 @@ class SubquerySuite extends QueryTest with SharedSQLContext {
+    checkAnswer(
+      sql("select l.a from l where (select sum(r.d) is null from r where l.a = r.c)"),
+      Row(1) :: Row(1) ::Row(null) :: Row(null) :: Row(6) :: Nil)
+
```
(Maybe remove these blank lines, https://github.com/databricks/scala-style-guide#blanklines)

frreiss replied:
Will do.

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala:
```diff
@@ -293,4 +293,65 @@ class SubquerySuite extends QueryTest with SharedSQLContext {
+      sql("select l.a from l where (select count(*) + sum(r.d) from r where l.a = r.c) = 0"),
+      Nil)
+
+
```
(Maybe here too..)

rxin commented on sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala:
```diff
@@ -293,4 +293,65 @@ class SubquerySuite extends QueryTest with SharedSQLContext {
       Row(3) :: Nil)
   }
+
+  test("COUNT bug in WHERE clause (Filter)") {
```
can you add the jira ticket number to the test case names?

frreiss replied:
Will do.

rxin commented:
cc @hvanhovell @davies

SparkQA commented:
**[Test build #2990 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2990/consoleFull)** for PR 13155 at commit [`76d9f45`](https://github.com/apache/spark/commit/76d9f4528b8536d1e5680279ab76b9e26dd3a873).

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1648,16 +1648,56 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
       subqueries: ArrayBuffer[ScalarSubquery]): LogicalPlan = {
     subqueries.foldLeft(child) {
       case (currentChild, ScalarSubquery(query, conditions, _)) =>
+        val aggOutputExpr = query.asInstanceOf[Aggregate].aggregateExpressions.head
```
You can actually get the aggregate in the `case ...` statement

hvanhovell replied:
I am also not sure that you should assume that an aggregate is returned.

hvanhovell replied:
For example, the (somewhat contrived) query:
```scala
range(0, 10).registerTempTable("a")
val plan = sql("select * from a where (select count(*) x from a aa where aa.id = a.id and aa.id % 2 = 0 having x > 0) > 0")
plan.explain(true)
```
results in the following analyzed plan:
```
== Analyzed Logical Plan ==
id: bigint
Project [id#0L]
+- Filter (scalar-subquery#27 [(id#0L#41L = id#0L)] > cast(0 as bigint))
   :  +- SubqueryAlias scalar-subquery#27 [(id#0L#41L = id#0L)]
   :     +- Project [x#26L,id#0L AS id#0L#41L]
   :        +- Filter (x#26L > cast(0 as bigint))
   :           +- Aggregate [id#0L], [(count(1),mode=Complete,isDistinct=false) AS x#26L,id#0L]
   :              +- Filter ((id#0L % cast(2 as bigint)) = cast(0 as bigint))
   :                 +- SubqueryAlias aa
   :                    +- SubqueryAlias a
   :                       +- Range 0, 10, 1, 8, [id#0L]
   +- SubqueryAlias a
      +- Range 0, 10, 1, 8, [id#0L]
```

frreiss replied:
Interesting. I had thought that the checks in the Analyzer would ensure that only Aggregate nodes could appear immediately under a ScalarSubquery node. Obviously that is not the case.

frreiss replied:
I'm looking into exactly what types of plans can show up in a ScalarSubquery node that passes all the checks upstream of RewriteCorrelatedScalarSubquery. Should have an answer in a few hours.

hvanhovell replied:
I can help you with that. The following construction 'rules' apply:

1. A `ScalarSubquery` has to contain an `Aggregate`. The Aggregate is can be nested in a `Filter` when a `HAVING` clauses is specified. This is the basic `ScalarSubquery`.
2. If the `ScalarSubquery` produces an `Attribute` that is also present in the outer query, then we wrap it with a `Project` and alias the offending attributes in order to prevent attribute clashes between the inner and outer queries.
3. In case of SQL generation we can add further `Project` and `SubqueryAlias` nodes.

The example applies rules 1 & 2.

frreiss replied:
Update: The relevant upstream checks are in CheckAnalysis.checkAnalysis():
```scala
case s @ ScalarSubquery(query, conditions, _) if conditions.nonEmpty =>
    ...
    def cleanQuery(p: LogicalPlan): LogicalPlan = p match {
      case SubqueryAlias(_, child) => cleanQuery(child)
      case Project(_, child) => cleanQuery(child)
      case child => child
    }

    cleanQuery(query) match {
      case a: Aggregate => checkAggregate(a)
      case Filter(_, a: Aggregate) => checkAggregate(a)
      case fail => failAnalysis(s"Correlated scalar subqueries must be Aggregated: $fail")
    }
```
These checks will accept a chain of operators above the Aggregate that match
the regex:
    (SubqueryAlias|Project)*(Filter)?

It doesn't look like AstBuilder will generate a SubqueryAlias nested immediately inside a SubqueryAlias, but all other matches of the above regex are possible.

I'll update `evalOnZeroTups()` above to rewrite combinations of Project, Filter, and Aggregate into an expression, then statically evaluate the expression.

frreiss replied:
Sorry, didn't see your reply before I posted mine. I must not have refreshed my browser. Thanks for the info on the possible cases. I'm testing the updated static evaluation code now.

frreiss replied:
Upon further reflection, a Filter node above the Aggregate creates additional issues. If you rewrite the subquery into an outer join above the Filter, it may be impossible to distinguish between two cases:
- The original subquery returns null because an aggregate value did not pass the HAVING clause
- The aggregate in the original subquery is evaluated over zero tuples

In both cases, the outer join will return a tuple containing nulls.
For example, the query:
```sql
select * from l where (select count(*) cnt from r where l.a = r.c having cnt = 0) = 0
```
would turn into
```sql
select * 
from l left outer join (select c, count(*) cnt from r group by c having cnt = 0) sq
    on l.a = sq.c 
where sq.cnt = 0
```
Note how the rewritten subquery `select c, count(*) cnt from r group by c having cnt = 0` always returns zero tuples.

I'm working through a few possible solutions, trying to see if one of them is guaranteed to be correct.

hvanhovell replied:
We could move the filter into the `left outer join` condition; this is not trivial but might work.

frreiss replied:
I came across an additional counterexample just now:
```sql
select l.a from l where
        (select case when count(*) = 1 then null else count(*) from r where l.a = r.c) = 0
```
This subquery returns null when count(*) on the inner query block is 1. The rewrite in this PR turns the overall query into:
```sql
select * 
from l left outer join 
    (select c, case when count(*) = 1 then null else count(*) as cnt from r group by c) sq
    on l.a = sq.c 
where coalesce(sq.cnt, 0) = 0
```
This result is incorrect; if exactly one tuple from R joins with a tuple of L, the query will return the L tuple, even though that tuple should not be returned.

So, to summarize, the operators above the outer join need to be able to discern between *four* different cases:
1. A tuple from the outer query block joins with one or more tuples in the subquery and produces an aggregate result that is *not* null
2. A tuple from the outer query block joins with one or more tuples in the subquery and produces an aggregate result that *is* null
3. A tuple from the outer query block does not join with any tuples in the subquery
4. A tuple from the outer query block joins with tuples in the subquery and produces an aggregate result that is filtered out by a HAVING clause

A single result column is simply not able to encode all four cases (or even the first three). The output of the outer join needs to have at least two columns to pass through enough information. 
Here's a scheme that will work as far as I can see: The result of the outer join contains the correlation columns, plus a column `aggVal` for the aggregate value, plus an additional column `isFiltered`. The four cases above are encoded as follows:
1. `aggVal is not null and isFiltered = false`
2. `aggVal is null and isFiltered = false`
3. `isFiltered is null`
4. `isFiltered = true`

Then the coalesce expression above the outer join turns into a CASE statement:
```sql
case 
   when isFiltered is null then coalesce(aggVal, <SQ return value on empty input>)
   when isFiltered then null
   else aggVal
```

Here's pseudocode for a rewrite that should produce the correct values of `isFiltered` and `aggVal`:
```
if (subquery returns null when zero tuples join) 
then
  Use the original rewrite.
else 
  if (subquery has a Filter above the Aggregate node)
  then
    Replace the Filter node with a Project that computes the value of isFiltered.
    Rewrite nodes above the Filter node so they pass through the isFiltered column.
  else
     Add an isFiltered column with a hard-coded value of false to the top Project in the subquery.
  endif
  Create a left outer join between the outer query block and the rewritten subquery.
  Put the case statement in the previous listing into the Project operator above the outer join.
endif
```

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1648,16 +1648,56 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+    // in the expression with the value they would return for zero input tuples.
+    val rewrittenExpr = expr transform {
+      case a @ AggregateExpression(aggFunc, _, _, resultId) =>
+        val resultLit = aggFunc.defaultResult match {
```
`getOrElse(...)?`

frreiss replied:
Works for me. I'll change that expression to getOrElse().

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1648,16 +1648,56 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+          case Some(lit) => lit
+          case None => Literal.default(NullType)
+        }
+        Alias(resultLit, "aggVal") (exprId = resultId)
```
Nit: space

hvanhovell replied:
Why do we need to use an `Alias` here? We are only evaluating the expression, and are not binding to the result of another operator.

frreiss replied:
The Alias was paranoia on my part. I noticed that AggregateExpression nodes had exprIds and I wasn't sure if those IDs could be consumed somewhere else in the expression evaluation code, so I added the Alias to retain the original IDs. All my test cases work without that additional Alias. I'll remove the Alias node.

SparkQA commented:
**[Test build #2990 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2990/consoleFull)** for PR 13155 at commit [`76d9f45`](https://github.com/apache/spark/commit/76d9f4528b8536d1e5680279ab76b9e26dd3a873).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1648,16 +1648,56 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+
+            // Renumber the original output, because the outer query refers to its ID.
+            val newQuery = query transformExpressions {
+              case Alias(c, n) => Alias(c, n)(exprId = newExprId)
```
This will give every alias the same id! That is not what we want.

frreiss replied:
Oops, there should be an "if" after that "case" expression. I'll update that expression to only renumber `origExprId` to `newExprId`.

hvanhovell commented:
@frreiss thanks for working on this. The fix you are proposing is the right one. There are still a few cornercases which need to be addressed before we can go ahead and merge this; see the comments.

frreiss committed:
Second version of the updated rewrite

frreiss committed:
Merge branch 'master' of https://github.com/apache/spark

frreiss commented:
I've added additional changes to cover two additional cases that @hvanhovell pointed out on review, plus one additional case that came up while fixing the first two:
- The correlated subquery may have a HAVING clause
- The correlated subquery may be nested inside additional query blocks that apply projections
- The correlated subquery may return NULL when the correlation bindings join
  with the subquery and a non-NULL value when bindings do *not* join, e.g. 
```sql
select l.a from l where
  (select case when count(*) = 1 then null else count(*) end from r where l.a = r.c) = 0
```

The logic for statically evaluating the subquery's aggregate expression now handles an Aggregate node with a chain of Filter and Project operators above it. The rewrite logic has an third case to handle subqueries with HAVING clauses. The second case from the original change set now adds an additional column to cover subqueries that return NULL with bindings that join the inner query block. I also addressed various other more minor review comments.

After these changes, the algorithm for preventing the COUNT bug in scalar subqueries is now the following:
```
V <- <value that the subquery returns when zero tuples join>
if (V is null)
then
  Use the original rewrite from SPARK-14785.
else 
  if (subquery has a Filter above the Aggregate node)
  then
    Rewrite the Filter node to a Project that adds a Boolean column isFiltered.
    Rewrite nodes above the Filter node so they pass through the isFiltered column.
  else
    Add an isFiltered column with a hard-coded value of false to the top Project in the subquery.
  endif
  Create a left outer join between the outer query block and the rewritten subquery.
  Put the following case statement into the Project operator above the outer join.
      case 
       when isFiltered is null then coalesce(aggVal, V)
       when isFiltered then null
       else aggVal
endif
```

### Correctness proof 

Let *b* denote a tuple of correlation bindings from the outer query block.  Without loss of generality, assume that the correlated subquery has the plan `Project(Filter(Aggregate(Join({b},T))))`, where *T* is a table. Note that upstream checks in Catalyst ensure that the Aggregate node will produce exactly one tuple.

Consider the evaluation of the original subquery on *b*. We can ask three questions about this evaluation:
- Did *b* join with one or more tuples from *T*?
- Did the Filter node reject the tuple that the Aggregate node returned?
- Did the subquery return `null`?

The answer to each of these questions must be "yes" or "no", leading to eight cases:

Case \# | Empty join? | Agg filtered? | SQ returns null?
------ | ----------- | ------------- | ----------------
1      | Yes         | Yes           | Yes
2      | Yes         | Yes           | No
3      | Yes         | No            | Yes
4      | Yes         | No            | No
5      | No          | Yes           | Yes
6      | No          | Yes           | No
7      | No          | No            | Yes
8      | No          | No            | No

Cases 2 and 6 are impossible. For the remaining cases, the rewritten subquery returns the correct result:
- Case 1 and 3 => Subquery returns null on empty join result, so first branch of rewrite algorithm above applies. Outer join returns null for subquery result.
- Case 4 => Subquery returns non-null answer on empty join result, so second or third branch of rewrite applies. Outer join in the rewritten query returns a tuple with `(aggVal, isFiltered)` set to `(null,null)`, so case statement at the top of the rewritten query returns the answer the subquery returns on an empty join, which is the correct answer.
- Case 5 => If first branch of rewrite applies, the outer join in the rewritten query returns null. If second or third branch of rewrite applies, then outer join in rewritten query returns a tuple with `(aggVal, isFiltered)` set to `(null, true)`, so case statement at top of rewritten query returns null.
- Case 7 => If first branch of rewrite applies, the outer join in the rewritten query returns null. For second or third branch of rewrite, outer join in rewritten query returns a tuple with `(aggVal, isFiltered)` set to `(null, false)`, so case statement at top of rewritten query returns null.
- Case 8 => If first branch of rewrite applies, the outer join in the rewritten query returns null. For second or third branch of rewrite, outer join in rewritten query returns a tuple with `(aggVal, isFiltered)` set to `(<non-null value>, false)`, so case statement at top of rewritten query returns the non-null value of aggVal.

frreiss commented:
Could one of the committers please trigger another build on this PR? The change set passes all the tests on my machine, but it's good to be safe.

SparkQA commented:
**[Test build #3013 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3013/consoleFull)** for PR 13155 at commit [`1b4ba5e`](https://github.com/apache/spark/commit/1b4ba5ed629d9b1e72d919d89b3592f7b29f3f3c).

SparkQA commented:
**[Test build #3014 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3014/consoleFull)** for PR 13155 at commit [`1b4ba5e`](https://github.com/apache/spark/commit/1b4ba5ed629d9b1e72d919d89b3592f7b29f3f3c).

SparkQA commented:
**[Test build #3013 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3013/consoleFull)** for PR 13155 at commit [`1b4ba5e`](https://github.com/apache/spark/commit/1b4ba5ed629d9b1e72d919d89b3592f7b29f3f3c).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

frreiss committed:
Merge branch 'master' of https://github.com/apache/spark

SparkQA commented:
**[Test build #3014 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3014/consoleFull)** for PR 13155 at commit [`1b4ba5e`](https://github.com/apache/spark/commit/1b4ba5ed629d9b1e72d919d89b3592f7b29f3f3c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

hvanhovell commented:
@frreiss this looks promising. I will take a more in depth look at it tomorrow evening (GMT).

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1695,176 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+   * Statically evaluate an expression containing zero or more placeholders, given a set
+   * of bindings for placeholder values.
+   */
+  private def evalExpr(expr : Expression, bindings : Map[Long, Option[Any]]) : Option[Any] = {
```
no space before ':'

frreiss replied:
Fixed that in my local copy. Will include the change in the next update.

hvanhovell replied:
Why do you use `ExprId.id` instead of entire `ExprId` as map keys?

frreiss replied:
No particular reason for using `ExprId.id`. I'll change to using the entire class as a key.

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1695,176 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+          // Renumber the original output, because the outer query refers to its ID.
+          val newExprId = NamedExpression.newExprId
+          val renumberedQuery = query transformExpressions {
+            case a@Alias(c, n) if a.exprId == origOutput.exprId => Alias(c, n)(exprId = newExprId)
```
Is it possible that there is origOutput.exprId come from an AttributeReference? for example:
```
SELECT (SELECT gid FROM g where gid=t.id) FROM t
```

frreiss replied:
I wasn't able to make the output of the subquery be an anonymous AttributeReference in my testing. The Analyzer adds aliases when the subquery's schema is anonymous.

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1695,176 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+          //      ELSE aggVal END
+          val caseExpr = Alias(CaseWhen(
+            Seq((IsNull(isFilteredRef), Coalesce(Seq(aggValRef,
+              Literal(resultWithZeroTups.getOrElse(null))))),
```
resultWithZeroTups can't be empty here, so it's better to use `resultWithZeroTups.get`

frreiss replied:
Fixed in my local copy.

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1695,176 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+            Join(currentChild, query, LeftOuter, conditions.reduceOption(And)))
+        } else {
+          // Renumber the original output, because the outer query refers to its ID.
+          val newExprId = NamedExpression.newExprId
```
It's a little bit risk to replace all the exprId with new one, the exprId could be cached somewhere without replacing.

Maybe we just use the same exprId for new Alias. It's a little bit confusing, but we already do this in a few places (change the nullability of a column without creating new one).

davies replied:
Another option could be update the expressions in outer query to use the new exprId (by keep these mapping as an hash map).

frreiss replied:
I originally tried using the same exprID in the new Alias, but a downstream rewrite kept going into infinite recursion. Not sure which one -- the stack kept blowing up before the trace said which rule was running. I figured that multiple uses of the same exprID across operators was not supported. If that approach is supposed to work, I can track down the rule that was blowing up and open another defect if you'd like.

frreiss replied:
Update: The infinite recursion was due to two instances of using `transform` instead of `transformUp` when replacing Alias nodes with their referents. Fixed those in my local copy and got rid of the ID renumbering in the rule.

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1695,176 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+   * @param filteredId expression ID for the "isFiltered" column
+   */
+  private def addIsFiltered(subqueryPlan : LogicalPlan, filteredId : ExprId) : LogicalPlan = {
+    val isFilteredRef = AttributeReference("isFiltered", BooleanType)(exprId = filteredId)
```
Why create a new one, but not passing the `isFilteredRef ` here?

frreiss replied:
This code no longer exists in my local copy after the refactoring below.

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1695,176 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+      // Correlated subquery without HAVING clause
+      // Add an additional Project that adds a constant value for "isFiltered"
+      rewrittenQuery = Project(subqueryPlan.output :+ Alias(Literal(false), "isFiltered")
+      (exprId = filteredId), subqueryPlan)
```
style: use a variable for `isFiltered `?

frreiss replied:
Good idea. Added a constant in my local copy.

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1695,176 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+  private def addIsFiltered(subqueryPlan : LogicalPlan, filteredId : ExprId) : LogicalPlan = {
+    val isFilteredRef = AttributeReference("isFiltered", BooleanType)(exprId = filteredId)
+    val (topPart, aggNode) = splitSubquery(subqueryPlan)
+    var rewrittenQuery: LogicalPlan = null
```
```
if () {
  var rewrittenQuery: LogicalPlan = filterAsProject
   while() {
   }
  rewrittenQuery
} else {
  Project(xx, subqueryPlan)
}
```

frreiss replied:
Fixed in my copy.

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1695,176 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+
+          val filteredId = NamedExpression.newExprId
+          val isFilteredRef = AttributeReference("isFiltered", BooleanType)(exprId = filteredId)
+          val withIsFiltered = addIsFiltered(renumberedQuery, filteredId)
```
Instead of turn the predicates in subquery into a projection, can we pull out the predicate in HAVING, then evaluate that after the join, so the result column will be.
```
Project(currentChild.output :+ If(predicate, agg, Literal(null)),
  Project(currentChild.output :+ Alias(Coalesce(Seq(aggvalRef, resultWithZeroTups)), "agg")())
```

frreiss replied:
I thought about pulling the filter expression up, but projections above the innermost clause of the subquery can modify the values that the HAVING clause references. For example, the subquery could be something like:
```
(select cnt + 1 from (select count(*) as cnt from r where r.a = s.b having cnt >= s.c))
```
To handle these cases correctly, we would need to rewrite all of the projection expressions to pass through any columns used by the HAVING clause; either that or selectively pull up the HAVING clause based on a static analysis of the projections above the innermost correlated subquery.

davies replied:
I think we could keep the columns that is used by the predicate. Above Aggregate, there should be only Project and Filter (optional), should be easy, right?

frreiss replied:
Sure, I can modify the rewrite there to pull the rewritten filter above the join.

hvanhovell replied:
@frreiss we already pull-out correlated predicates in the Analyzer. Adding these predicates might simplify this rule.

davies commented:
@frreiss Thanks for working on this. Had left some comments on how to rewrite the subquery, let me know how do you think, thanks.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1695,176 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+          if (bindings.size == 0) bindings
+          else {
+            val exprResult = evalExpr(condition, bindings).getOrElse(false)
+              .asInstanceOf[Boolean]
```
Do you need the type cast?

frreiss replied:
The test on the next line doesn't compile without the cast. The condition in a Filter node is of type Expression, and `Expresssion.eval()` returns `Any`.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1695,176 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+        case SubqueryAlias(_, child) => evalPlan(child)
+        case Filter(condition, child) =>
+          val bindings = evalPlan(child)
+          if (bindings.size == 0) bindings
```
NIT: `bindings.isEmpty`?

frreiss replied:
Fixed in my local copy

frreiss replied:
Fixed in my local copy.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1695,176 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+          else {
+            val exprResult = evalExpr(condition, bindings).getOrElse(false)
+              .asInstanceOf[Boolean]
+            if (exprResult) bindings else Map()
```
Nit: `Map.empty`

frreiss replied:
Fixed in my local copy

frreiss replied:
Fixed in my local copy.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1695,176 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+  private def evalAggOnZeroTups(expr : Expression) : Option[Any] = {
+    // AggregateExpressions are Unevaluable, so we need to replace all aggregates
+    // in the expression with the value they would return for zero input tuples.
+    val rewrittenExpr = expr transform {
```
I think we also need to `null` out `AttributeReference`s. This is a (contrived) example when we need this:
```scala
range(0, 15).createOrReplaceTempView("a")
range(0, 100).select((rand() * 10).cast("int").as("key"), $"id".as("value")).createOrReplaceTempView("b")
sql(
"""
select  a.id,
        (select b.key + count(*) from b where b.key = a.id) cnt
from    a
""")
```

frreiss replied:
Yes, that is true. Added a test case and an additional clause in that case statement in my local copy.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1695,176 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+          // for joining with the outer query block. Fill those expressions in with
+          // nulls and statically evaluate the remainder.
+          aggExprs.map(ne => ne match {
+            case AttributeReference(_, _, _, _) => (ne.exprId.id, None)
```
We could just call `evalAggOnZeroTups` on every expression. That would also allow for more complex groupBy expressions (currently not needed though).

frreiss replied:
The join attributes that the Analyzer adds to the Aggregate node are AttributeReference nodes, and the `evalAggOnZeroTups` method as currently written can't process them. A more comprehensive facility for statically evaluating expressions would be nice to have; but I'm hesitant to add such a mechanism as part of a bug fix. Perhaps a follow-on JIRA is in order?

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1695,176 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+   * (first part of returned value) and the parts below the Aggregate node, including
+   * the Aggregate (second part of returned value)
+   */
+  private def splitSubquery(plan : LogicalPlan) : Tuple2[Seq[LogicalPlan], Aggregate] = {
```
Use `(Seq[LogicalPlan], Aggregate)` instead of `Tuple2[...]`

frreiss replied:
Fixed in my local copy.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1695,176 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+   * the Aggregate (second part of returned value)
+   */
+  private def splitSubquery(plan : LogicalPlan) : Tuple2[Seq[LogicalPlan], Aggregate] = {
+    var topPart = List[LogicalPlan]()
```
Use an ArrayBuffer here.

frreiss replied:
Fixed in my local copy.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1695,176 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+          Project(
+            currentChild.output :+ origOutput,
+            Join(currentChild, query, LeftOuter, conditions.reduceOption(And)))
+        } else {
```
It might be easier to check if the first column is `nullable`. Then we only need to do the `evalSubqueryOnZeroTups` eval if we need to, and then we might have an opportunity to simplify the code.

frreiss replied:
Sorry, I'm having some trouble understanding this comment. I think you're suggesting that the rule should first check whether `query.output.head` (the output column of the scalar subquery) is a nullable column; then the rule should sometimes skip the call to `evalQueryOnZeroTups()` depending on the nullability of the column.

I don't see how the check for nullability would allow one to skip the call to `evalQueryOnZeroTups()`. If the column is nullable, then the correlated subquery could return either null or a non-null value when zero tuples match; and the rule needs to call `evalQueryOnZeroTups` to distinguish between those two cases. If the column is not nullable, then the rule still needs to call `evalQueryOnZeroTups` to determine what non-null value the subquery returns when zero tuples match.

hvanhovell replied:
@frreiss this bug only occurs when the result of the scalar subquery cannot be null. `COUNT(...)` being the prime example of this. We can identify this problem early and easily by checking the nullable flag. Am I missing something?

BTW: Nice talk on SystemML!

frreiss replied:
@hvanhovell the conditions for the COUNT bug in this rule are actually a bit less strict. If the correlated subquery returns an answer other than NULL when the correlation bindings do not join, then it may be possible to have the COUNT bug. Take this query for example:
```sql
select l.a from l where
        (select case when count(*) = 1 then null else count(*) end as cnt
        from r where l.a = r.c) = 0
```
This subquery returns NULL when exactly 1 tuple from `r` joins with a set of correlation bindings and a non-NULL value otherwise. Rewrite this query without compensating for the COUNT bug and you get:
```sql
select l.a from l left join 
  (select r.c, case when count(*) = 1 then null else count(*) end as cnt from r group by r.c) SQ
on l.a = SQ.c
where SQ.cnt = 0
```
If the value of `l.a` matches zero tuples in `r`, the outer join produces a tuple with `cnt` set to NULL and the rewritten query incorrectly returns an empty result.

Technically the full condition for the COUNT bug is that the subquery returns a non-NULL value when zero tuples join; *and* the outer query block will return a different answer if it receives that value instead of NULL. Checking the outer query block would be a somewhat-useful performance optimization, but it would require additional static evaluation code.

frreiss commented:
Thanks @hvanhovell for the additional pass of review! I'll be preparing my slides for Spark Summit all day today but will come back to this PR as soon as that's done.

rxin commented:
@frreiss  did you have a chance to update this? We want to cut an RC asap.

frreiss commented:
@rxin I'll have an updated set of changes in tonight

frreiss committed:
Addressing additional corner cases and review comments.

frreiss commented:
Updated changes are in. Running a full regression suite overnight.

frreiss commented:
Tests ran successfully on my machine.

SparkQA commented:
**[Test build #3074 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3074/consoleFull)** for PR 13155 at commit [`8cd2877`](https://github.com/apache/spark/commit/8cd2877179dded4557c8da92e5b16011637289b0).

SparkQA commented:
**[Test build #3074 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3074/consoleFull)** for PR 13155 at commit [`8cd2877`](https://github.com/apache/spark/commit/8cd2877179dded4557c8da92e5b16011637289b0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

davies commented:
LGTM

davies commented:
@hvanhovell Could you have a final look on this?

hvanhovell commented:
@frreiss PR looks good. It does need a little bit of work in the style department.  I'll leave comments. Lemme know if you have time to address these; else I'll take over and do that chore (you'll still get the credit though).

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1696,205 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+      case a @ AggregateExpression(aggFunc, _, _, resultId) =>
+        aggFunc.defaultResult.getOrElse(Literal.default(NullType))
+
+      case AttributeReference(_, _, _, _) => Literal.default(NullType)
```
style: `case _: AttributeReference => ...`

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1696,205 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+    // Inputs to this method will start with a chain of zero or more SubqueryAlias
+    // and Project operators, followed by an optional Filter, followed by an
+    // Aggregate. Traverse the operators recursively.
+    def evalPlan(lp : LogicalPlan) : Map[ExprId, Option[Any]] = {
```
style: no space before colon

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1696,205 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+          // Some of the expressions under the Aggregate node are the join columns
+          // for joining with the outer query block. Fill those expressions in with
+          // nulls and statically evaluate the remainder.
+          aggExprs.map(ne => ne match {
```
style: do the pattern match directly, i.e.:
```scala
aggExpr.map {
  case _: AttributeReference =>
  case ...
}
```

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1696,205 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+          // nulls and statically evaluate the remainder.
+          aggExprs.map(ne => ne match {
+            case AttributeReference(_, _, _, _) => (ne.exprId, None)
+            case Alias(AttributeReference(_, _, _, _), _) => (ne.exprId, None)
```
style: `case Alias(_: AttributeReference, _) =>`

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1696,205 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+   */
+  private def splitSubquery(plan: LogicalPlan) : (Seq[LogicalPlan], Option[Filter], Aggregate) = {
+    val topPart = ArrayBuffer.empty[LogicalPlan]
+    var bottomPart : LogicalPlan = plan
```
style: no space before colon

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1696,205 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+    var bottomPart : LogicalPlan = plan
+    while (true) {
+      bottomPart match {
+        case havingPart@Filter(_, aggPart@Aggregate(_, _, _)) =>
```
style spaces between @ elements, i.e.: `havingPart @ Filter`

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1696,205 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+            // CASE 3: Subquery with HAVING clause. Pull the HAVING clause above the join.
+            // Need to modify any operators below the join to pass through all columns
+            // referenced in the HAVING clause.
+            var subqueryRoot : UnaryNode = aggNode
```
Style colon. 
same for line below

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1696,205 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+            var subqueryRoot : UnaryNode = aggNode
+            val havingInputs : Seq[NamedExpression] = aggNode.output
+
+            topPart.reverse.foreach(
```
style. do the pattern match in the foreach, i.e.:
```scala
topPart.reverse.foreach {
  case Project(projList, _) =>
  ...
}
```

Also @'s

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1695,16 +1696,205 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+            //      WHEN NOT (original HAVING clause expr) THEN CAST(null AS <type of aggVal>)
+            //      ELSE (aggregate value) END AS (original column name)
+            val caseExpr = Alias(CaseWhen(
+              Seq[(Expression, Expression)] (
```
Do we need to type the Seq?

hvanhovell commented:
LGTM - merging to master/2.0 (will do a small follow-up)

sameeragarwal commented:
@hvanhovell  maybe this broke the build? https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Compile/job/spark-branch-2.0-compile-maven-scala-2.10/310/
